File: applicationsOfMobileAndroid/app/src/main/java/com/crm369/mobile/MobileNotificationStateStore.kt
package com.appswebnetkz.crm369

import android.content.Context

class MobileNotificationStateStore(context: Context) {
    private val preferences = context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)

    fun newEntries(
        domainKey: String,
        feed: MobileNotificationFeed,
    ): List<MobileNotificationEntry> {
        if (feed.userId <= 0) {
            return emptyList()
        }

        val identity = identity(domainKey, feed.userId)
        val currentIdentity = preferences.getString(KEY_IDENTITY, null)
        val deliveredKeys = if (currentIdentity == identity) {
            preferences.getStringSet(KEY_DELIVERED_KEYS, emptySet()).orEmpty()
        } else {
            emptySet()
        }

        return feed.entries.filterNot { deliveredKeys.contains(it.key) }
    }

    fun markDelivered(
        domainKey: String,
        feed: MobileNotificationFeed,
        deliveredKeys: List<String>,
    ) {
        if (feed.userId <= 0 || deliveredKeys.isEmpty()) {
            return
        }

        val identity = identity(domainKey, feed.userId)
        val currentIdentity = preferences.getString(KEY_IDENTITY, null)
        val existingKeys = if (currentIdentity == identity) {
            preferences.getStringSet(KEY_DELIVERED_KEYS, emptySet()).orEmpty().toList()
        } else {
            emptyList()
        }

        val mergedKeys = (deliveredKeys + existingKeys)
            .distinct()
            .take(MAX_TRACKED_KEYS)
            .toSet()

        preferences.edit()
            .putString(KEY_IDENTITY, identity)
            .putStringSet(KEY_DELIVERED_KEYS, mergedKeys)
            .apply()
    }

    private fun identity(domainKey: String, userId: Int): String {
        return "$domainKey|$userId"
    }

    companion object {
        private const val PREFERENCES_NAME = "crm369_mobile_notifications"
        private const val KEY_IDENTITY = "identity"
        private const val KEY_DELIVERED_KEYS = "delivered_keys"
        private const val MAX_TRACKED_KEYS = 200
    }
}

File: applicationsOfMobileIos/CRM369iOS/MobileNotificationStateStore.swift
import Foundation

final class MobileNotificationStateStore {
    private let defaults: UserDefaults
    private let identityKey = "crm369.notifications.identity"
    private let deliveredKeysKey = "crm369.notifications.delivered_keys"
    private let maximumTrackedKeys = 200

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func newEntries(baseURL: URL, feed: MobileNotificationFeed) -> [MobileNotificationEntry] {
        guard feed.userID > 0 else {
            return []
        }

        let feedIdentity = identity(baseURL: baseURL, userID: feed.userID)
        let deliveredKeys = defaults.string(forKey: identityKey) == feedIdentity
            ? Set(defaults.stringArray(forKey: deliveredKeysKey) ?? [])
            : []

        return feed.entries.filter { !deliveredKeys.contains($0.key) }
    }

    func markDelivered(baseURL: URL, feed: MobileNotificationFeed, keys: [String]) {
        guard feed.userID > 0, !keys.isEmpty else {
            return
        }

        let feedIdentity = identity(baseURL: baseURL, userID: feed.userID)
        let existingKeys = defaults.string(forKey: identityKey) == feedIdentity
            ? defaults.stringArray(forKey: deliveredKeysKey) ?? []
            : []
        let mergedKeys = Array(NSOrderedSet(array: keys + existingKeys))
            .compactMap { $0 as? String }
            .prefix(maximumTrackedKeys)

        defaults.set(feedIdentity, forKey: identityKey)
        defaults.set(Array(mergedKeys), forKey: deliveredKeysKey)
    }

    private func identity(baseURL: URL, userID: Int) -> String {
        "\(baseURL.absoluteString)|\(userID)"
    }
}
